[css-scoping] Implement `:has-slotted` pseudo-class

keithamus opened the pull request:
#### 91135a212d4d93f0cb68154a91d2b11303378c95
<pre>
[css-scoping] Implement `:has-slotted` pseudo-class
<a href="https://bugs.webkit.org/show_bug.cgi?id=280608">https://bugs.webkit.org/show_bug.cgi?id=280608</a>

Reviewed by NOBODY (OOPS!).

This implements the :has-slotted pseudo selector as specified in CSSWG
<a href="https://drafts.csswg.org/css-scoping/#the-has-slotted-pseudo.">https://drafts.csswg.org/css-scoping/#the-has-slotted-pseudo.</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-changing-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-changing-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-manual-assignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-query-selector-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing/parse-has-slotted.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesHasSlottedPseudoClass):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::setManuallyAssignedSlot):
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didChangeSlot):
(WebCore::NamedSlotAssignment::willRemoveAssignedNode):
(WebCore::NamedSlotAssignment::assignToSlot):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::containsAssignedNodeIgnoringSlots):
(WebCore::HTMLSlotElement::hasFlattenedSlottedContent const):
* Source/WebCore/html/HTMLSlotElement.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91135a212d4d93f0cb68154a91d2b11303378c95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19848 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115781 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59994 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30096 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38006 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83411 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112708 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23983 "Found 1 new test failure: fast/filter-image/clipped-filter.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63874 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23363 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16998 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59575 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/102251 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118573 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108312 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36799 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27259 "Found 1 new test failure: fast/shadow-dom/host-child-append-performance.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92416 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37172 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92238 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37204 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14950 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32628 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36694 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42164 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132588 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36354 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35893 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39696 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -154,7 +155,8 @@ void add(Hasher& hasher, const CSSParserContext& context)
         | context.cssProgressFunctionEnabled                << 28
-        | (uint32_t)context.mode                            << 29; // This is multiple bits, so keep it last.
+        | context.cssHasSlottedEnabled                      << 29
+        | (uint32_t)context.mode                            << 30; // This is multiple bits, so keep it last.
```
CSSParserMode has 5 values and requires 3 bits so this won't work. We probably need to get rid of one of the existing flags in order to do this.

rniwa replied:
Maybe we should get rid of customStateSetEnabled. It has been stable & shipping for a while.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -385,6 +386,9 @@ void NamedSlotAssignment::willRemoveAssignedNode(const Node& node, ShadowRoot&)
     });
+
+    if (auto* slotElement = slot->element.get())
+        Style::PseudoClassChangeInvalidation styleInvalidation(*slotElement, CSSSelector::PseudoClass::HasSlotted, !slot->assignedNodes.isEmpty());
```
We need to create this object before we remove the matching items in the lines above and destroy it after we removed the items so something like this:
```cpp
std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
if (auto* slotElement = slot->element.get())
    styleInvalidation.emplace(*slotElement, CSSSelector::PseudoClass::HasSlotted, !slot->assignedNodes.isEmpty());
slot->assignedNodes.removeFirstMatching([&node](const auto& item) {
    return item.get() == &node;
});

```

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -416,24 +420,33 @@ void NamedSlotAssignment::assignSlots(ShadowRoot& shadowRoot)
-    for (auto& entry : m_slots)
+    for (auto& entry : m_slots) {
         entry.value->assignedNodes.shrinkToFit();
+        if (auto* slotElement = entry.value->element.get())
```
Ditto.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -416,24 +420,33 @@ void NamedSlotAssignment::assignSlots(ShadowRoot& shadowRoot)
+        if (defaultSlotEntry != m_slots.end()) {
             defaultSlotEntry->value->assignedNodes.append(child);
+            if (auto* slotElement = defaultSlotEntry->value->element.get())
+                Style::PseudoClassChangeInvalidation styleInvalidation(*slotElement, CSSSelector::PseudoClass::HasSlotted, true);
```
Ditto.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -416,24 +420,33 @@ void NamedSlotAssignment::assignSlots(ShadowRoot& shadowRoot)
     addResult.iterator->value->assignedNodes.append(child);
+
+    if (auto* slotElement = addResult.iterator->value->element.get())
+        Style::PseudoClassChangeInvalidation styleInvalidation(*slotElement, CSSSelector::PseudoClass::HasSlotted, true);
```
Ditto.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-001.tentative-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL Changing content targetting :has-slotted(...) and :has-slotted through a single shadow root assert_equals: expected "rgb(0, 0,  255)" but got "rgb(0, 255, 0)"
+FAIL Changing content targetting :has-slotted(...) and :has-slotted through a single shadow root assert_equals: expected "rgb(0, 255, 0)" but got "rgb(0, 0, 255)"
```
Do we understand why this is still failing?

keithamus replied:
This tests is for the functional (tentative) variant of `:has-slotted()`, vs what this PR implements which is the non-functional `:has-slotted`.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-003.tentative-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL Changing content targetting :has-slotted(...) and :has-slotted through multiple shadow roots assert_equals: expected "rgb(0, 0,  255)" but got "rgb(0, 255, 0)"
+FAIL Changing content targetting :has-slotted(...) and :has-slotted through multiple shadow roots assert_equals: expected "rgb(0, 255, 0)" but got "rgb(0, 0, 255)"
```
Do we understand why this is still failing?

keithamus replied:
Likewise, this test is for `:has-slotted()` which this PR doesn't implement.

darinadler commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -263,6 +263,12 @@ ALWAYS_INLINE bool matchesDirPseudoClass(const Element& element, const AtomStrin
+ALWAYS_INLINE bool matchesHasSlottedPseudoClass(const Element& element)
+{
+    auto* slotElement = dynamicDowncast<HTMLSlotElement>(element);
+    return slotElement ? slotElement->hasFlattenedSlottedContent() : false;
```
```suggestion
    return slotElement && slotElement->hasFlattenedSlottedContent();
```

darinadler commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -157,7 +158,8 @@ void add(Hasher& hasher, const CSSParserContext& context)
         | context.cssRandomFunctionEnabled                  << 27
-        | (uint32_t)context.mode                            << 28; // This is multiple bits, so keep it last.
+        | context.cssHasSlottedEnabled                      << 28
+        | (uint32_t)context.mode                            << 29; // This is multiple bits, so keep it last.
```
Is 2 bits enough? Maybe the comment could specify the number of bits? Or we could have a static assertion that confirms the number of bits.

rniwa replied:
We have 5 values in this enum so 2 bits isn't enough:
```cpp
enum CSSParserMode : uint8_t {
    HTMLStandardMode,
    HTMLQuirksMode,
    // SVG attributes are parsed in quirks mode but rules differ slightly.
    SVGAttributeMode,
    // User agent stylesheets are parsed in standards mode but also allows internal properties and values.
    UASheetMode,
    // WebVTT places limitations on external resources.
    WebVTTMode
};
```

rniwa replied:
We should add an assert somewhere to validate that every value of CSSParserMode can be stored & loaded properly.

darinadler commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -387,6 +388,10 @@ void NamedSlotAssignment::willRemoveAssignedNode(const Node& node, ShadowRoot&)
+
+    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
+    if (auto* slotElement = slot->element.get())
+        styleInvalidation.emplace(*slotElement, CSSSelector::PseudoClass::HasSlotted, slotElement->hasFlattenedSlottedContent());
```
Is there a way to share more of this code? This sequence of three lines of code is pretty repetitive, shows up 5 times almost identically, and I think a helper function could improve clarity.

nt1m replied:
`Style::PseudoClassChangeInvalidation` is scope-dependent. It would be possible to define another scope that wraps this scope, but not worth it IMO

darinadler commented on Source/WebCore/html/HTMLSlotElement.cpp:
```diff
@@ -145,6 +145,13 @@ static void flattenAssignedNodes(Vector<Ref<Node>>& nodes, const HTMLSlotElement
+{
+    Vector<Ref<Node>> nodes;
+    flattenAssignedNodes(nodes, *this);
+    return !nodes.isEmpty();
```
Seems inefficient to build the list of nodes just to check if it’s empty. Can we do it more efficiently?

darinadler reviewed (approved):
I think this is OK as is.

rniwa reviewed (changes requested):
We need more than 1 bits for `context.mode`.

keithamus committed:
[css-scoping] Implement `:has-slotted` pseudo-class
https://bugs.webkit.org/show_bug.cgi?id=280608

Reviewed by NOBODY (OOPS!).

This implements the :has-slotted pseudo selector as specified in CSSWG
https://drafts.csswg.org/css-scoping/#the-has-slotted-pseudo.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-changing-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-changing-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-001.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-functional-changing-003.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-manual-assignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scoping/has-slotted-query-selector-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing/parse-has-slotted.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesHasSlottedPseudoClass):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::setManuallyAssignedSlot):
* Source/WebCore/dom/SlotAssignment.cpp:
(WebCore::NamedSlotAssignment::didChangeSlot):
(WebCore::NamedSlotAssignment::willRemoveAssignedNode):
(WebCore::NamedSlotAssignment::assignToSlot):
* Source/WebCore/html/HTMLSlotElement.cpp:
(WebCore::containsAssignedNodeIgnoringSlots):
(WebCore::HTMLSlotElement::hasFlattenedSlottedContent const):
* Source/WebCore/html/HTMLSlotElement.h:

rniwa commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -112,6 +112,7 @@ CSSParserContext::CSSParserContext(const Document& document, const URL& sheetBas
     , cssDynamicRangeLimitMixEnabled { document.settings().cssDynamicRangeLimitMixEnabled() }
     , cssConstrainedDynamicRangeLimitEnabled { document.settings().cssConstrainedDynamicRangeLimitEnabled() }
+    , cssHasSlottedEnabled { document.settings().cssHasSlottedEnabled() }
     , webkitMediaTextTrackDisplayQuirkEnabled { document.quirks().needsWebKitMediaTextTrackDisplayQuirk() }
```
Why is this value not being included in the hash computation?

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1375,8 +1376,17 @@ void Node::setManuallyAssignedSlot(HTMLSlotElement* slotElement)
     else if (RefPtr text = dynamicDowncast<Text>(*this))
         RenderTreeUpdater::tearDownRenderer(*text);
 
+    auto* oldSlotElement = manuallyAssignedSlot();
```
Please use `RefPtr` instead.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -311,9 +312,14 @@ void NamedSlotAssignment::didChangeSlot(const AtomString& slotAttrValue, ShadowR
     auto assignedNodes = std::exchange(slot->assignedNodes, { });
     m_slotAssignmentsIsValid = false;
 
+    if (auto* slotElement = slot->element.get())
```
Please use `findFirstSlotElement` and `RefPtr` instead.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -390,10 +396,15 @@ void NamedSlotAssignment::willRemoveAssignedNode(Node& node, ShadowRoot&)
         return item.get() == &node;
     });
 
+    if (auto* slotElement = slot->element.get())
```
Ditto.

rniwa commented on Source/WebCore/dom/SlotAssignment.cpp:
```diff
@@ -441,15 +452,21 @@ void NamedSlotAssignment::assignSlots(ShadowRoot& shadowRoot)
+        if (defaultSlotEntry != m_slots.end()) {
             defaultSlotEntry->value->assignedNodes.append(child);
+            if (auto* slotElement = defaultSlotEntry->value->element.get())
+                styleInvalidation.emplace(*slotElement, CSSSelector::PseudoClass::HasSlotted, slotElement->hasFlattenedSlottedContent());
```
This is too late. `NamedSlotAssignment::assignSlots` is only called lazily.
We need to detect when `:has-slotted` is used and eagerly resolve the assignment.
See how `shadowRoot.shouldFireSlotchangeEvent() && hasAssignedNodes(shadowRoot, slot)` is checked to enqueue `slotchange` event.

rniwa commented on Source/WebCore/html/HTMLSlotElement.cpp:
```diff
@@ -147,6 +147,43 @@ static void flattenAssignedNodes(Vector<Ref<Node>>& nodes, const HTMLSlotElement
+    auto* assignedNodes = slot.assignedNodes();
+    if (!assignedNodes) {
+        for (RefPtr<Node> child = slot.firstChild(); child; child = child->nextSibling()) {
+            if (auto* slot = dynamicDowncast<HTMLSlotElement>(*child)) {
```
Please use RefPtr.

rniwa commented on Source/WebCore/html/HTMLSlotElement.cpp:
```diff
@@ -147,6 +147,43 @@ static void flattenAssignedNodes(Vector<Ref<Node>>& nodes, const HTMLSlotElement
+            ASSERT_NOT_REACHED();
+            continue;
+        }
+        if (auto* slot = dynamicDowncast<HTMLSlotElement>(*weakNode)) {
```
Ditto.

rniwa commented on Source/WebCore/html/HTMLSlotElement.h:
```diff
@@ -56,6 +56,9 @@ class HTMLSlotElement final : public HTMLElement {
 
     void updateAccessibilityOnSlotChange() const;
+
+    bool hasFlattenedSlottedContent() const;
```
I'd add this code right beneath `assignedElements` instead.
